test/chain-prune-test.js

```diff
@@ -0,0 +1,201 @@
+/* eslint-env mocha */
+/* eslint prefer-arrow-callback: "off" */
+
+'use strict';
+
+const assert = require('bsert');
+const Network = require('../lib/protocol/network');
+const WorkerPool = require('../lib/workers/workerpool');
+const Miner = require('../lib/mining/miner');
+const Chain = require('../lib/blockchain/chain');
+
+const network = Network.get('regtest');
+
+describe('Chain Prune', function() {
+  describe('Prune options', function() {
+    const chainOptions = {
+      memory: true,
+      network
+    };
+
+    it('should not allow retroactive prune', async () => {
+      const chain = new Chain(chainOptions);
+
+      await chain.open();
+      await chain.close();
+
+      // Retroactively enable prune.
+      await assert.rejects(async () => {
+        chain.options.prune = true;
+        await chain.open();
+      }, {
+        message: 'Cannot retroactively prune.'
+      });
+      await chain.close();
+    });
+
+    it('should not allow retroactive unprune', async () => {
+      const chain = new Chain({
+        ...chainOptions,
+        prune: true
+      });
+
+      await chain.open();
+      await chain.close();
+
+      await assert.rejects(async () => {
+        chain.options.prune = false;
+        await chain.open();
+      }, {
+        message: 'Cannot retroactively unprune.'
+      });
+      await chain.close();
+    });
+
+    it('should force retroactive prune', async () => {
+      const chain = new Chain({
+        ...chainOptions
+      });
+
+      await chain.open();
+      await chain.close();
+
+      let pruned = false;
+      await assert.doesNotReject(async () => {
+        chain.options.forceFlags = true;
+        chain.options.prune = true;
+
+        chain.db.prune = () => {
+          pruned = true;
+        };
+
+        await chain.open();
+      });
+
+      await chain.close();
+      assert.strictEqual(pruned, true, 'should call prune.');
+    });
+  });
+
+  describe('Prune', function() {
+    const PRUNE_AFTER_HEIGHT = network.block.pruneAfterHeight;
+    const KEEP_BLOCKS = network.block.keepBlocks;
+
+    const workers = new WorkerPool({
+      enabled: true,
+      size: 2
+    });
+
+    const chainoptions = {
+      memory: true,
+      network,
+      workers
+    };
+
+    let chain, miner, cpu;
+
+    beforeEach(async () => {
+      await workers.open();
+
+      network.block.pruneAfterHeight = 10;
+      network.block.keepBlocks = 10;
+
+      chain = new Chain(chainoptions);
+      miner = new Miner({ chain });
+      cpu = miner.cpu;
+
+      await miner.open();
+    });
+
+    afterEach(async () => {
+      network.block.pruneAfterHeight = PRUNE_AFTER_HEIGHT;
+      network.block.keepBlocks = KEEP_BLOCKS;
+
+      if (chain.opened)
+        await chain.close();
+
+      await miner.close();
+      await workers.open();
```

Comment:
should this be `workers.close()`? (this is the `afterEach`)

Reply:
yes it should, nice catch.